src/dune_pkg/rev_store.ml

```diff
@@ -84,18 +138,21 @@ let show =
 ;;
 
 let load_or_create ~dir =
-  let t = { dir } in
+  let parent = dir |> Path.parent_exn in
+  let lock = Path.relative parent "rev-store.lock" in
+  let t = { dir; lock } in
   let* () = Fiber.return () in
   let+ () =
-    match Fpath.mkdir_p (Path.to_string dir) with
-    | Already_exists -> Fiber.return ()
-    | Created -> run t [ "init"; "--bare" ]
-    | exception Unix.Unix_error (e, x, y) ->
-      User_error.raise
-        [ Pp.textf "%s isn't a directory" (Path.to_string_maybe_quoted dir)
-        ; Pp.textf "reason: %s" (Unix_error.Detailed.to_string_hum (e, x, y))
-        ]
-        ~hints:[ Pp.text "delete this file or check its permissions" ]
+    with_flock lock ~f:(fun () ->
+      match Fpath.mkdir_p (Path.to_string dir) with
+      | Already_exists -> Fiber.return ()
+      | Created -> run t [ "init"; "--bare" ]
+      | exception Unix.Unix_error (e, x, y) ->
```

Comment:
`(e, x, y)` are meaningless names and this code just passes the tuple on to `Unix_error.Detailed.to_string_hum` so I'd suggest not destructuring the tuple here.

Reply:
I thought so too, but in this case `Unix_error` is an constructor of arity 3, and not a constructor of arity 1 with a tuple, so it can't be bound to a tuple value:

```ocaml
# type a = A of (int * string) | B of int * string;;
# let x = A (3, "42");;
val x : a = A (3, "42")
# let x = B (3, "42");;
val x : a = B (3, "42")
# match x with | A v -> () | B v -> ();;
Error: The constructor B expects 2 argument(s),
       but is applied here to 1 argument(s)
```